src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -353,40 +359,39 @@ where
         _serializer: &mut EventSerializer,
         _start: u64,
     ) -> Result<SinkReply> {
-        let client = self.client.as_mut().ok_or(ErrorKind::ClientNotAvailable(
-            "BigQuery",
-            "The client is not connected",
-        ))?;
-        let write_stream = self
-            .write_stream
-            .as_ref()
-            .ok_or(ErrorKind::ClientNotAvailable(
-                "BigQuery",
-                "The write stream is not available",
-            ))?;
-        let mapping = self.mapping.as_mut().ok_or(ErrorKind::ClientNotAvailable(
-            "BigQuery",
-            "The mapping is not available",
-        ))?;
-
-        let mut request_data = Vec::with_capacity(1);
-
-        let mut request = AppendRowsRequest {
-            write_stream: write_stream.name.clone(),
-            offset: None,
-            rows: Some(append_rows_request::Rows::ProtoRows(ProtoData {
-                writer_schema: Some(ProtoSchema {
-                    proto_descriptor: Some(mapping.descriptor().clone()),
-                }),
-                rows: Some(ProtoRows {
-                    serialized_rows: Vec::with_capacity(event.len()),
-                }),
-            })),
-            trace_id: String::new(),
-        };
-
-        for data in event.value_iter() {
-            let serialized_event = mapping.map(data)?;
+        let mut request_data: Vec<AppendRowsRequest> = Vec::new();
+
+        let mut requests: HashMap<String, AppendRowsRequest> = HashMap::new();
+
+        for (data, meta) in event.value_meta_iter() {
+            let write_stream = self
+                .get_or_create_write_stream(
+                    &ctx.extract_meta(meta)
+                        .get("table_id")
+                        .as_str()
+                        .map_or_else(|| self.config.table_id.clone(), ToString::to_string),
+                    ctx,
+                )
+                .await?;
+
+            let mut request =
```

Comment:
I think instead of repeatedly adding and removing the [Entry API](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html) with `or_insert` would do a better job there.